### PR TITLE
Build multiple databases in travis ci

### PR DESCRIPTION
**Pull request by malakai97**
This pull request adds support for building sqlite, mysql, and postgres in travis.  

Three files were created to enable this: 
- config/database.yml.travis-mysql
- config/database.yml.travis-sqlite
- config/database.yml.travis-postgres

Additionally .travis.yml was extended to move the above files into place as well as setup the build matrix.

No changes to the ruby/rails code were involved.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,20 +1,44 @@
...
+  - DB=postgres
+
+before_install:
+  - if [[ $DB = 'sqlite' ]]; then echo "gem 'sqlite3'" > Gemfile.local; fi
+  - if [[ $DB = 'mysql' ]]; then echo "gem 'mysql2'" > Gemfile.local; fi
+  - if [[ $DB = 'postgres' ]]; then echo "gem 'pg'" > Gemfile.local; fi
+
 install:
   - bundle install --jobs=3 --retry=3
-  - bundle exec sqlite3 --version
```
Comment (dazza-codes) [inline .travis.yml:11]: Could we keep this and add similar db `--version` calls, using a similar conditional pattern to the `before_install` block?  We've already run into some version specific problems in migrations, so we're going to want these sooner or later so we know what travis is running (unless we get that info some other way?).

File: .travis.yml [hunk 2]
```diff
@@ -1,20 +1,44 @@
...
+before_script:
   - bundle exec rake config
+  - cp config/database.yml.travis-$DB config/database.yml
+  - if [[ $DB = 'sqlite' ]]; then bundle exec rake db:create; fi
+  - if [[ $DB = 'mysql' ]]; then mysql -e 'CREATE DATABASE dpn;'; fi
+  - if [[ $DB = 'postgres' ]]; then psql -c 'CREATE DATABASE dpn;' -U postgres; fi
+  - bundle exec rake db:migrate
+
+script: bundle exec rspec --tag ~broken_in_ci
+
```
Comment (dazza-codes) [inline .travis.yml:39]: The MySQL and Postgres builds will likely work without the `--tag ~broken_in_ci`.  We could add more `if $DB` clauses here to only exclude stuff for sqlite3 or we could change that tag so it's `broken_in_sqlite`?

File: config/database.yml.travis-postgres [hunk 3]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2015 The Regents of the University of Michigan.
+# All Rights Reserved.
+# Licensed according to the terms of the Revised BSD License
+# See LICENSE.md for details.
+
+test: 
+  adapter: postgresql
+  database: dpn
```
Comment (dazza-codes) [inline config/database.yml.travis-postgres:8]: Seems too easy - don't need to specify a user and passwd?

File: .travis.yml [hunk 4]
```diff
@@ -1,20 +1,44 @@
...
-script: bundle exec rspec --tag ~broken_in_ci
+
+services:
+  - mysql
+  - postgresql
+
+env:
+  - DB=sqlite
+  - DB=mysql
+  - DB=postgres
```
Comment (dazza-codes) [inline .travis.yml:17]: This bit is confusing.  It seems to set them all and I'm not sure how this works out, maybe the last one wins or something?

File: .travis.yml [hunk 5]
```diff
@@ -1,20 +1,44 @@
...
+
 branches:
   only:
     - master
     - develop
-script: bundle exec rspec --tag ~broken_in_ci
+
+services:
+  - mysql
+  - postgresql
```
Comment (dazza-codes) [inline .travis.yml:12]: Can this also use the `env` value for `$DB` to define the service required?

File: .travis.yml [hunk 6]
```diff
@@ -1,20 +1,44 @@
...
 install:
   - bundle install --jobs=3 --retry=3
-  - bundle exec sqlite3 --version
+
+before_script:
   - bundle exec rake config
+  - cp config/database.yml.travis-$DB config/database.yml
+  - if [[ $DB = 'sqlite' ]]; then bundle exec rake db:create; fi
+  - if [[ $DB = 'mysql' ]]; then mysql -e 'CREATE DATABASE dpn;'; fi
+  - if [[ $DB = 'postgres' ]]; then psql -c 'CREATE DATABASE dpn;' -U postgres; fi
```
Comment (dazza-codes) [inline .travis.yml:35]: Wondering if this `psql` needs to set the owner of the db?  On my laptop, I had to use something like

```
createdb -T template0 --owner=dpnAdmin dpn_development
```

Maybe that's only required when the user is not `postgres`.

**Review by dazza-codes — COMMENTED** [review]
Looks pretty good, just a few questions on how things work and whether we can keep the db `--version` stuff.  Also, would be good to drop the `broken_in_ci` tag for MySQL and Postgres.

File: config/database.yml.travis-postgres (hunk 3 above)
Comment (malakai97) [reply config/database.yml.travis-postgres:8]: https://docs.travis-ci.com/user/database-setup/#PostgreSQL

File: .travis.yml (hunk 5 above)
Comment (malakai97) [reply .travis.yml:12]: If there is, it's not documented.  I tried a couple of the obvious things and couldn't get it to work.

File: .travis.yml (hunk 4 above)
Comment (malakai97) [reply .travis.yml:17]: https://docs.travis-ci.com/user/customizing-the-build

File: .travis.yml (hunk 6 above)
Comment (malakai97) [reply .travis.yml:35]: /shrug I just did what the documentation said to do: https://docs.travis-ci.com/user/database-setup/#PostgreSQL

File: .travis.yml (hunk 1 above)
Comment (malakai97) [reply .travis.yml:11]: Indeed; I completely forgot that I removed this.

File: .travis.yml (hunk 2 above)
Comment (malakai97) [reply .travis.yml:39]: I want to be very careful about the complexity in the ci server leaking into the codebase.  Having the extra database.yml files was a necessary evil I couldn't manage to workaround.  So restricting `broken_in_ci` to `broken_in_sqlite` is great, but I don't want to one day also have `broken_in_mysql`, `broken_in_postgres`, `broken_in_mongo`, `broken_on_tuesdays`, `broken_in_linux`, etc.

Comment (dazza-codes) [reply .travis.yml:39]: OK, let's just go with `broken` ;-)

**Comment by dazza-codes** [comment]
Some postgres errors in https://travis-ci.org/dpn-admin/dpn-server/jobs/170304591 that might be ActiveRecord bugs or might require restructuring the AR method calls?

Not sure if this is actually a blocker for this PR, because the intent of this PR is actually fulfilled by running the Postgres tests.  Do we need to fix this stuff for this PR?

**Comment by dazza-codes** [comment]
I'm going to make a call on this and say this PR is doing what it should - it builds against all 3 db platforms intended.

**Pull request merged**